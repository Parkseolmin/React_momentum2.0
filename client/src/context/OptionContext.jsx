import { createContext, useContext, useEffect, useRef, useState } from 'react';

const OptionContext = createContext();

export function OptionProvider({ children }) {
  const [showItemBox, setShowItemBox] = useState(false);
  const itemBoxRef = useRef();
  const clockRef = useRef();

  useEffect(() => {
    function handleClickOutside(event) {
      if (
        itemBoxRef.current &&
        !itemBoxRef.current.contains(event.target) &&
        clockRef.current &&
        !clockRef.current.contains(event.target)
      ) {
        setShowItemBox(false);
      }
    }
    document.addEventListener('mousedown', handleClickOutside);
    return () => document.removeEventListener('mousedown', handleClickOutside);
  }, [itemBoxRef, clockRef]);

  return (
    <OptionContext.Provider
      value={{ showItemBox, setShowItemBox, itemBoxRef, clockRef }}
    >
      {children}
    </OptionContext.Provider>
  );
}

export const useOptionContext = () => useContext(OptionContext);
