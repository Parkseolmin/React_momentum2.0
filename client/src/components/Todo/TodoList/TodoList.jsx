import { useState } from 'react';
import styles from './TodoList.module.css';
import AddTodo from '../AddTodo/AddTodo';
import TodoItem from '../TodoItem/TodoItem';
import { useTodos } from 'context/TodosContext';

export default function TodoList({ filter }) {
  const [searchTerm, setSearchTerm] = useState('');
  const { todos, dispatch } = useTodos();

  const handleAdd = (todo) => {
    dispatch({ type: 'ADD', todo });
  };

  const handleUpdate = (updated) => {
    dispatch({ type: 'UPDATE', updated });
  };

  const handleDelete = (deleted) => {
    dispatch({ type: 'DELETE', deleted });
  };

  const handleSearch = (e) => {
    setSearchTerm(e.target.value);
  };

  function getFilteredItems(filter, searchTerm) {
    if (filter === 'all') {
      if (!searchTerm) {
        return todos;
      } else {
        return todos.filter((todo) =>
          todo.text.toLowerCase().includes(searchTerm.toLowerCase())
        );
      }
    } else {
      if (!searchTerm) {
        return todos.filter((todo) => todo.status === filter);
      } else {
        return todos
          .filter((todo) => todo.status === filter)
          .filter((todo) =>
            todo.text.toLowerCase().includes(searchTerm.toLowerCase())
          );
      }
    }
  }

  const filtered = getFilteredItems(filter, searchTerm);

  return (
    <section className={styles.container}>
      <input
        className={styles.searchInput}
        type='text'
        placeholder='검색...'
        value={searchTerm}
        onChange={handleSearch}
      />
      <ul className={styles.list}>
        {filtered.map((item) => {
          return (
            <TodoItem
              key={item.id}
              todo={item}
              onUpdate={handleUpdate}
              onDelete={handleDelete}
            />
          );
        })}
      </ul>
      <AddTodo onAdd={handleAdd} />
    </section>
  );
}
