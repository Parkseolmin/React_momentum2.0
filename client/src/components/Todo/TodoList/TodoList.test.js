import { fireEvent, render, screen } from '@testing-library/react';
import TodoList from './TodoList';
import { TodosProvider } from 'context/TodosContext';

describe('TodoList Component', () => {
  beforeAll(() => {
    window.HTMLElement.prototype.scrollIntoView = jest.fn();
  });

  afterAll(() => {
    delete window.HTMLElement.prototype.scrollIntoView;
  });

  it('Snapshot Test', () => {
    const { asFragment } = render(
      <TodosProvider>
        <TodoList filter='all' />
      </TodosProvider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders essential components correctly', () => {
    render(
      <TodosProvider>
        <TodoList filter='all' />
      </TodosProvider>
    );

    expect(screen.getByPlaceholderText('검색...')).toBeInTheDocument();
    expect(
      screen.getByRole('button', { name: /todoAddButton/i })
    ).toBeInTheDocument();
    expect(screen.getByRole('list')).toBeInTheDocument();
  });

  it('adds a new todo to the list when the add button is clicked', () => {
    render(
      <TodosProvider>
        <TodoList filter='all' />
      </TodosProvider>
    );

    const input = screen.getByPlaceholderText('오늘 할 일');
    const addButton = screen.getByRole('button', { name: /todoAddButton/i });

    fireEvent.change(input, { target: { value: 'Learn Testing' } });
    expect(input.value).toBe('Learn Testing');

    fireEvent.click(addButton);
    expect(screen.getByText('Learn Testing')).toBeInTheDocument();
  });

  describe('ADD, UPDATEM, DELETE', () => {
    it('dispatched ADD action with the correct payload', () => {
      const mockDispatch = jest.fn();
      jest.spyOn(require('context/TodosContext'), 'useTodos').mockReturnValue({
        todos: [],
        dispatch: mockDispatch,
      });

      render(
        <TodosProvider>
          <TodoList filter='all' />
        </TodosProvider>
      );

      const input = screen.getByPlaceholderText('오늘 할 일');
      const addButton = screen.getByRole('button', { name: /todoAddButton/i });

      fireEvent.change(input, { target: { value: 'Learn Testing' } });
      fireEvent.click(addButton);

      expect(mockDispatch).toHaveBeenCalledWith({
        type: 'ADD',
        todo: expect.objectContaining({
          text: 'Learn Testing',
          status: 'active',
        }),
      });
    });

    it('should toggle the status of a todo when checkbox is clicked', () => {
      const mockTodos = [{ id: '1', text: 'Learn Testing', status: 'active' }];

      const mockDispatch = jest.fn();
      jest.spyOn(require('context/TodosContext'), 'useTodos').mockReturnValue({
        todos: mockTodos,
        dispatch: mockDispatch,
      });

      render(
        <TodosProvider>
          <TodoList filter='all' />
        </TodosProvider>
      );

      const checkbox = screen.getByRole('checkbox', { name: /Learn Testing/i });
      fireEvent.click(checkbox);

      expect(mockDispatch).toHaveBeenCalledWith({
        type: 'UPDATE',
        updated: { ...mockTodos[0], status: 'completed' },
      });
    });

    it('should delete a todo when delete button is clicked', () => {
      const mockTodos = [{ id: '1', text: 'Learn Testing', status: 'active' }];

      const mockDispatch = jest.fn();
      jest.spyOn(require('context/TodosContext'), 'useTodos').mockReturnValue({
        todos: mockTodos,
        dispatch: mockDispatch,
      });

      render(
        <TodosProvider>
          <TodoList filter='all' />
        </TodosProvider>
      );

      const deleteButton = screen.getByRole('button', {
        name: /todoDeleteButton/i,
      });
      fireEvent.click(deleteButton);

      expect(mockDispatch).toHaveBeenCalledWith({
        type: 'DELETE',
        deleted: mockTodos[0],
      });
    });
  });

  describe('filrers method', () => {
    it('filters todos based on search term', () => {
      // Mock todos before rendering
      const mockTodos = [
        { id: 1, text: 'Learn React', status: 'active' },
        { id: 2, text: 'Learn Testing', status: 'completed' },
        { id: 3, text: 'Write Tests', status: 'active' },
      ];

      jest.spyOn(require('context/TodosContext'), 'useTodos').mockReturnValue({
        todos: mockTodos,
        dispatch: jest.fn(),
      });

      render(
        <TodosProvider>
          <TodoList filter='all' />
        </TodosProvider>
      );

      const input = screen.getByPlaceholderText('검색...');
      fireEvent.change(input, { target: { value: 'Learn' } });

      // Verify filtered todos
      expect(screen.getByText('Learn React')).toBeInTheDocument();
      expect(screen.getByText('Learn Testing')).toBeInTheDocument();
      expect(screen.queryByText('Write Tests')).not.toBeInTheDocument();
    });

    it('filters todos based on status when no search term is provided', () => {
      const mockTodos = [
        { id: 1, text: 'Learn React', status: 'active' },
        { id: 2, text: 'Learn Testing', status: 'completed' },
        { id: 3, text: 'Write Tests', status: 'active' },
      ];

      jest.spyOn(require('context/TodosContext'), 'useTodos').mockReturnValue({
        todos: mockTodos,
        dispatch: jest.fn(),
      });

      render(
        <TodosProvider>
          <TodoList filter='completed' />
        </TodosProvider>
      );

      // 'completed' 상태의 todo만 보여야 함
      expect(screen.getByText('Learn Testing')).toBeInTheDocument();
      expect(screen.queryByText('Learn React')).not.toBeInTheDocument();
      expect(screen.queryByText('Write Tests')).not.toBeInTheDocument();
    });

    it('filters todos based on status and search term', () => {
      const mockTodos = [
        { id: 1, text: 'Learn React', status: 'active' },
        { id: 2, text: 'Learn Testing', status: 'completed' },
        { id: 3, text: 'Write Tests', status: 'active' },
      ];

      jest.spyOn(require('context/TodosContext'), 'useTodos').mockReturnValue({
        todos: mockTodos,
        dispatch: jest.fn(),
      });

      render(
        <TodosProvider>
          <TodoList filter='active' />
        </TodosProvider>
      );

      const input = screen.getByPlaceholderText('검색...');
      fireEvent.change(input, { target: { value: 'Write' } });

      // 'active' 상태이면서 검색어 'Write'가 포함된 todo만 보여야 함
      expect(screen.getByText('Write Tests')).toBeInTheDocument();
      expect(screen.queryByText('Learn React')).not.toBeInTheDocument();
      expect(screen.queryByText('Learn Testing')).not.toBeInTheDocument();
    });
  });
  it('displays message when no todos are available', () => {
    jest.spyOn(require('context/TodosContext'), 'useTodos').mockReturnValue({
      todos: [],
      dispatch: jest.fn(),
    });

    render(
      <TodosProvider>
        <TodoList filter='all' />
      </TodosProvider>
    );

    expect(screen.getByPlaceholderText('오늘 할 일')).toBeInTheDocument();
  });
});
