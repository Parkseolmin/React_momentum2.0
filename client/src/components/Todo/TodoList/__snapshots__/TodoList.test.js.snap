// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList Component Snapshot Test 1`] = `
<DocumentFragment>
  <section
    class="container"
  >
    <input
      class="searchInput"
      placeholder="검색..."
      type="text"
      value=""
    />
    <ul
      class="list"
    />
    <form
      class="form"
    >
      <input
        class="input"
        placeholder="오늘 할 일"
        type="text"
        value=""
      />
      <button
        aria-label="todoAddButton"
        class="button"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
          />
        </svg>
      </button>
    </form>
  </section>
</DocumentFragment>
`;
